### PR TITLE
Mark datatypes as injective in recursive module approximations

### PR DESCRIPTION
**Pull request by garrigue**
Protecting against non-wellfounded injectivity inferences resulted in rejecting some valid recursive module definitions in #10028 .
This fixes that by marking the approximation of a datatype as injective.
This should go in 4.12 too.

### CODE REVIEW

File: typing/typedecl.ml [hunk 1]
```diff
@@ -403,7 +403,7 @@ let transl_declaration env sdecl (id, uid) =
         type_kind = kind;
         type_private = sdecl.ptype_private;
         type_manifest = man;
-        type_variance = Variance.unknown_signature ~arity;
+        type_variance = Variance.unknown_signature ~injective:false ~arity;
```
Comment (lpw25) [inline typing/typedecl.ml:406]: Not sure it matters, but it seems like you could fill this one in based on the kind

**Review by lpw25 — APPROVED** [review]
Looks good to me

File: typing/typedecl.ml (hunk 1 above)
Comment (garrigue) [reply typing/typedecl.ml:406]: Actually, this information is never used, as the variance is computed in the next pass.
So I chose to keep the original default here.

**Comment by garrigue** [comment]
Thanks for the review. I did not add a log message, since this is new in 4.12.

**Comment by Octachron** [comment]
Cherry-picked on 4.12 as aad63626405c7f29be2 .

**Pull request merged**